### PR TITLE
Add configuration files for Python pre-commit workflow

### PR DESCRIPTION
**Pull request by tyronevb**
Added configuration files for black, flake8 and pre-commit. These tools
are used to auto-format and check Python source files that are staged
for commit.

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[flake8]
+max-line-length = 120
+max-complexity = 15
+ignore = E203, E501, W503
```
[the remark this answers is not in the document]
Comment (tyronevb) [reply .flake8]: Good point. I have added comments to the flake8 configuration file explaining which rules are enabled and which are ignored.

File: .flake8 [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[flake8]
+max-line-length = 120
+max-complexity = 15
```
[the remark this answers is not in the document]
Comment (tyronevb) [reply .flake8:6]: It's based on the [McCabe Cyclomatic Complexity](https://dzone.com/articles/what-exactly-is-mccabe-cyclomatic-complexity). The default is 10, but I've read that it is often relaxed to 15 to allow for more complicated methods.

**Comment by tyronevb** [comment]
I've now also added support for pydocstyle and mypy checks to the pre-commit workflow.

File: .flake8 [hunk 3]
```diff
@@ -1,5 +1,22 @@
...
+
+# configure the maximum complexity to allow
+# 15 is recommended for more complex modules
 max-complexity = 15
-ignore = E203, E501, W503
-select = C,E,F,W,B
\ No newline at end of file
+
+# specify errors/checks to ignore
+# this list is tailored to work well with the black formatter
```
Comment (james-smith-za) [inline .flake8:9]: I'm sure this must be a racist comment.

File: .flake8 (hunk 2 above)
[the remark this answers is not in the document]
Comment (gcallanan) [reply .flake8:6]: I like this - right up until it tells me my method is too complex. Then I will be unhappy.

**Pull request merged**